services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,572 @@
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbCallExecutor;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDocumentBuilder;
+import org.eclipse.hono.deviceregistry.mongodb.utils.MongoDbDeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.DeviceRegistryUtils;
+import org.eclipse.hono.deviceregistry.util.Versioned;
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.eclipse.hono.service.management.Result;
+import org.eclipse.hono.service.management.tenant.Tenant;
+import org.eclipse.hono.service.management.tenant.TenantManagementService;
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import io.opentracing.Span;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+import io.vertx.ext.mongo.MongoClientUpdateResult;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+public final class MongoDbBasedTenantService extends AbstractVerticle
+        implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private static final int INDEX_CREATION_MAX_RETRIES = 3;
+    private MongoClient mongoClient;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an instance of the {@code MongoDbCallExecutor} and {@code MongoClient}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    public MongoDbBasedTenantsConfigProperties getConfig() {
```

Comment:
This method could be removed and we could use directly use the instance variable `config`.

Reply:
done